### PR TITLE
Disable event before issuing orders to tradeship

### PR DESCRIPTION
**Pull request by Marsunpaisti**

### CODE REVIEW

File: wurst/systems/trade/TradeShipSupervisor.wurst [hunk 1]
```diff
@@ -44,20 +45,28 @@ public class TradeShipSupervisor
...
+        if GetIssuedOrderId() != Orders.smart and GetIssuedOrderId() != Orders.move
+            return
         switch state
             case MOVING
                 let orderId = GetIssuedOrderId()
                 let targetPos = getOrderPos()
                 let nextStopPos = nextStop.getPos()
                 if orderId != Orders.move or targetPos.distanceToSq(nextStopPos) > 10000.0
+                    this.orderListenerEnabled = false
                     ship.issuePointOrder("move", nextStopPos)
```
[the remark this answers is not in the document]
Comment (Marsunpaisti) [reply wurst/systems/trade/TradeShipSupervisor.wurst]: The event fires during the call. I had to do this trick in my AI troll experimentation as well

File: wurst/systems/trade/TradeShipSupervisor.wurst [hunk 2]
```diff
@@ -44,20 +45,28 @@ public class TradeShipSupervisor
...
-        EventListener.add(ship, EVENT_UNIT_ISSUED_POINT_ORDER, -> onIssuedOrder())
-        EventListener.add(ship, EVENT_UNIT_ISSUED_ORDER, -> onIssuedOrder())
+        EventListener.add(ship, EVENT_PLAYER_UNIT_ISSUED_TARGET_ORDER, -> onIssuedOrder())
+        EventListener.add(ship, EVENT_PLAYER_UNIT_ISSUED_POINT_ORDER, -> onIssuedOrder())
+        EventListener.add(ship, EVENT_PLAYER_UNIT_ISSUED_ORDER, -> onIssuedOrder())
 
     function onIssuedOrder()
+        if not this.orderListenerEnabled
+            return
+        if GetIssuedOrderId() != Orders.smart and GetIssuedOrderId() != Orders.move
```
Comment (jonathanmclaus) [inline wurst/systems/trade/TradeShipSupervisor.wurst]: What is Orders.smart used for?

File: wurst/systems/trade/TradeShipSupervisor.wurst [hunk 3]
```diff
@@ -44,20 +45,28 @@ public class TradeShipSupervisor
         nextStop = firstStop.getNext()
         ships.put(ship, this)
         moveToNextStop()
-        EventListener.add(ship, EVENT_UNIT_ISSUED_TARGET_ORDER, -> onIssuedOrder())
-        EventListener.add(ship, EVENT_UNIT_ISSUED_POINT_ORDER, -> onIssuedOrder())
-        EventListener.add(ship, EVENT_UNIT_ISSUED_ORDER, -> onIssuedOrder())
+        EventListener.add(ship, EVENT_PLAYER_UNIT_ISSUED_TARGET_ORDER, -> onIssuedOrder())
+        EventListener.add(ship, EVENT_PLAYER_UNIT_ISSUED_POINT_ORDER, -> onIssuedOrder())
+        EventListener.add(ship, EVENT_PLAYER_UNIT_ISSUED_ORDER, -> onIssuedOrder())
```
Comment (jonathanmclaus) [inline wurst/systems/trade/TradeShipSupervisor.wurst]: Which of these fires for what order? Do different ones fire for buying and actually issuing an order?

File: wurst/systems/trade/TradeShipSupervisor.wurst (hunk 2 above)
Comment (Marsunpaisti) [reply wurst/systems/trade/TradeShipSupervisor.wurst]: its the order given on a right click, having multiple actions depending on target

File: wurst/systems/trade/TradeShipSupervisor.wurst (hunk 3 above)
Comment (Marsunpaisti) [reply wurst/systems/trade/TradeShipSupervisor.wurst]: Depends on target type. I suppose buying is either an immediate order or target order. Move is a point order. Stop is an immediate order.

**Comment by jonathanmclaus** [comment]
How does this cause the issue where buying an item from a stopped boat would cause it move? Wouldn't its state be STOPPED so it just reissues the stop order?

File: wurst/systems/trade/TradeShipSupervisor.wurst [hunk 4]
```diff
@@ -75,11 +83,15 @@ public class TradeShipSupervisor
             return
 
         state = TradeShipSupervisorState.MOVING
+        this.orderListenerEnabled = false
         ship.issuePointOrder("move", nextStop.getPos())
+        this.orderListenerEnabled = true
         destroyCallback()
 
         callback = doPeriodically(1.0) cb ->
```
Comment (jonathanmclaus) [inline wurst/systems/trade/TradeShipSupervisor.wurst]: Do we need this anymore? I think it's only to override commands issued by players, which this blocks now.

Comment (Marsunpaisti) [reply wurst/systems/trade/TradeShipSupervisor.wurst]: I'd have the callback just to be safe

Comment (jonathanmclaus) [reply wurst/systems/trade/TradeShipSupervisor.wurst]: OK. Add a comment to mark it for possible deprecation and I'll accept. I don't want to let it fall through the cracks at least.

Comment (Marsunpaisti) [reply wurst/systems/trade/TradeShipSupervisor.wurst]: Done

**Pull request merged**